BCnEnc.Net/Decoder/BcBlockDecoder.cs

```diff
@@ -1,108 +1,131 @@
-        where T : unmanaged
-    {
-        public RawBlock4X4Rgba32[,] Decode(ReadOnlySpan<byte> data, int pixelWidth, int pixelHeight, out int blockWidth, out int blockHeight)
-        {
-            blockWidth = (int)MathF.Ceiling(pixelWidth / 4.0f);
-            blockHeight = (int)MathF.Ceiling(pixelHeight / 4.0f);
-
-            if (data.Length != blockWidth * blockHeight * Unsafe.SizeOf<T>())
-            {
-                throw new InvalidDataException("Given data does not match expected length.");
-            }
-
-            var encodedBlocks = MemoryMarshal.Cast<byte, T>(data);
-
-            var output = new RawBlock4X4Rgba32[blockWidth, blockHeight];
-
-            for (var x = 0; x < blockWidth; x++)
-            {
-                for (var y = 0; y < blockHeight; y++)
-                {
-                    output[x, y] = DecodeBlock(encodedBlocks[x + y * blockWidth]);
-                }
-            }
-
-            return output;
-        }
-
-        protected abstract RawBlock4X4Rgba32 DecodeBlock(T block);
-    }
-
-    internal class Bc1NoAlphaDecoder : BaseBcBlockDecoder<Bc1Block>
-    {
-        protected override RawBlock4X4Rgba32 DecodeBlock(Bc1Block block)
-        {
-            return block.Decode(false);
-        }
-    }
-
-    internal class Bc1ADecoder : BaseBcBlockDecoder<Bc1Block>
-    {
-        protected override RawBlock4X4Rgba32 DecodeBlock(Bc1Block block)
-        {
-            return block.Decode(true);
-        }
-    }
-
-    internal class Bc2Decoder : BaseBcBlockDecoder<Bc2Block>
-    {
-        protected override RawBlock4X4Rgba32 DecodeBlock(Bc2Block block)
-        {
-            return block.Decode();
-        }
-    }
-
-    internal class Bc3Decoder : BaseBcBlockDecoder<Bc3Block>
-    {
-        protected override RawBlock4X4Rgba32 DecodeBlock(Bc3Block block)
-        {
-            return block.Decode();
-        }
-    }
-
-    internal class Bc4Decoder : BaseBcBlockDecoder<Bc4Block>
-    {
-        private readonly bool redAsLuminance;
-
-        public Bc4Decoder(bool redAsLuminance)
-        {
-            this.redAsLuminance = redAsLuminance;
-        }
-
-        protected override RawBlock4X4Rgba32 DecodeBlock(Bc4Block block)
-        {
-            return block.Decode(redAsLuminance);
-        }
-    }
-
-    internal class Bc5Decoder : BaseBcBlockDecoder<Bc5Block>
-    {
-        protected override RawBlock4X4Rgba32 DecodeBlock(Bc5Block block)
-        {
-            return block.Decode();
-        }
-    }
-
-    internal class Bc7Decoder : BaseBcBlockDecoder<Bc7Block>
-    {
-        protected override RawBlock4X4Rgba32 DecodeBlock(Bc7Block block)
-        {
-            return block.Decode();
-        }
-    }
+	internal interface IBcBlockDecoder
+	{
+		RawBlock4X4Rgba32[,] Decode(ReadOnlyMemory<byte> data, int pixelWidth, int pixelHeight, OperationContext context,
+			out int blockWidth, out int blockHeight);
+	}
+
```

Comment:
Everything looks ok here.

Reply:
Does the method in the Base class look good as well? I'm not sure how much getting the Span from a ReadOnlyMemory<T> every block impacts performance. I can't pass an already existing Span to the delegate in Parallel.For due to it being a ref struct.

Reply:
It was not an issue at least in the encoder. Getting the span should be very fast. 

But since decoding is a reasonably fast operation, this might benefit from dividing the tasks to larger chunks (like  line by line).

Reply:
Encoding doesn't have this issue since it doesn't take in any Spans.
Decode on the other hand does, which makes it unusable for Parallel.For like this. It doesn't matter how we divide it.

If decoding is already that fast, maybe I shouldn't parallelism in decoding in general. Then decoding would only benefit from the cancel feature above.

Reply:
Ah, yeah I was thinking about the MemoryMarshal cast. This could changed to take byte[] since the public api doesn't take in Span/Memory anymore. But Memory.span is still very fast.

Could see how fast would it take to decode a large (4k) bc7 texture as the "worst case" scenario with and without parallelism. That would give an idea on the speedup gain from parallelization.

Reply:
Oh, so you would agree to changing it to a byte[]. That's a nice info. I thought about doing that, but I was sure there was a reason it was still used. Doesn't necessarily seem to be the case then. I think creating a span from an array or using Memory.Span should be equal in speed. So I would switch the decoding signature to consuming a byte[].

Reply:
It was using ReadOnlySpan\<byte\> before because the raw decoding function was taking in ReadOnlySpan\<byte\>. Span was used before to allow someone to have a long byte buffer with BC data in it somewhere and use the decoder with Span.Slice() without having to copy the compressed data part of the buffer to another array. Now that the RawDecoding has to take in a byte[] because of async limitations I think you can safely change this to take in byte[] as well.

Reply:
Hmm, I could possibly pass through a ReadOnlyMemory<byte> instead. It would enable slicing as well, and would have the same advantages as Span<T>. The only difference is performance drops with Memory<T> since it lives on the heap and does the same operations differently therefore.

So pass through a ReadOnlyMemory<T> from the start, and then call its Span property in the Decoding method like it already is here. Would you agree to that? Then this conversation can be resolved.

Reply:
That seems good.